ludwig/modules/lr_scheduler.py

```diff
@@ -0,0 +1,161 @@
+        self.step_info = StepInfo(steps_per_checkpoint, total_steps, self.config)
+        self._train_scheduler = get_schedule_with_warmup(self.config, self.optimizer, self.step_info)
+
+        # Scheduler updated each eval step
+        self._eval_scheduler = None
+        if self.config.reduce_on_plateau > 0:
+            mode = "min" if self.validation_metric.get_objective() == MINIMIZE else "max"
+            self._eval_scheduler = ReduceLROnPlateauLimited(
+                optimizer=self.optimizer,
+                mode=mode,
+                reduce_limit=self.config.reduce_on_plateau,
+                factor=self.config.reduce_on_plateau_rate,
+                patience=self.config.reduce_on_plateau_patience,
+            )
+
+        self.reset(steps_per_checkpoint, total_steps)
+
+    def reset(self, steps_per_checkpoint: int, total_steps: int):
+        # Retain state but update number of steps for training
+        self.step_info.reset(steps_per_checkpoint, total_steps)
+
+    def step(self):
+        self._train_scheduler.step()
+
+    def eval_step(self, progress_tracker: ProgressTracker, validation_field: str):
+        if self._eval_scheduler is None:
+            # No reduce on plateau
+            return
+
+        if self.config.reduce_eval_split == TRAINING:
+            split_metrics = progress_tracker.train_metrics
+        elif self.config.reduce_eval_split == VALIDATION:
+            split_metrics = progress_tracker.validation_metrics
+        else:  # if self.config.reduce_eval_split == TEST:
+            split_metrics = progress_tracker.test_metrics
+
+        validation_metric = self.config.reduce_eval_metric
+        last_metric: TrainerMetric = split_metrics[validation_field][validation_metric][-1]
+        last_metric_value = last_metric[-1]
+
+        self._eval_scheduler.step(last_metric_value)
+
+    def state_dict(self) -> Dict[str, Any]:
+        return {
+            "train_scheduler_state": self._train_scheduler.state_dict(),
+            "eval_scheduler_state": self._eval_scheduler.state_dict() if self._eval_scheduler is not None else {},
+        }
+
+    def load_state_dict(self, d: Dict[str, Any]):
+        self._train_scheduler.load_state_dict(d["train_scheduler_state"])
+        if self._eval_scheduler is not None:
+            self._eval_scheduler.load_state_dict(d["eval_scheduler_state"])
+
+
+class StepInfo:
+    def __init__(self, steps_per_checkpoint: int, total_steps: int, config: LRSchedulerConfig):
+        self.config = config
+        self.reset(steps_per_checkpoint, total_steps)
+
+    def reset(self, steps_per_checkpoint: int, total_steps: int):
+        self.steps_per_checkpoint = steps_per_checkpoint
+        self.num_training_steps = total_steps
+
+        if self.config.warmup_fraction > 0 and self.config.warmup_evaluations > 0:
+            logging.info(
+                "Both `learning_rate_scheduler.warmup_fraction` and `learning_rate_scheduler.warmup_evaluations`. "
```

Comment:
nit: Both warmups "are specified" or "are non-zero".

"The larger of the two (as a function of the total training steps) will be used."